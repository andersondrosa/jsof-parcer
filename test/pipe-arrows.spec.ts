import JSOF from "../dist";

const code1 = `
(x => x |> (
  (
    ((x, y) => ((x, y) |> sum)) |> ("calc" |> set),
    (x => x |> toUpper) |> ("upper" |> set),
    (m => m |> throw) |> ("Error" |> set),
    (() => (
      ("MockedFunctionJson" |> get) |> ("CreateFunctionByJson" |> get)
    )) |> ("composedFunctionToUpper" |> set),
    [ "alpha", "beta", "omega" ] |> ("data" |> set),
    (toUpper |> map) |> mock |> ("MockedFunctionJson" |> set),
    (json => json |> build) |> ("CreateFunctionByJson" |> set),
    "composedFunctionToUpper" |> get, 
    fn => (("data" |> get) |> fn) |> ("data" |> set),
    "example" |> ("upper" |> get) |> identity, 
    x => x |> toLower |> of,
    "data" |> get |> concat,
    ("TEST ERROR 1" |> ("Error" |> get), identity) |> try,
    ("acc", (acc, x) => (acc, (x |> toLower) |> of) |> concat) |> set,
    ("acc" |> get, []) |> reduce,
    (x => x |> toUpper ) |> map,
  ) 
  |> pipe
))
|> context
`;

test("must contain equal JSON", () => {
  expect(
    JSOF.parse(`
    context(x => pipe(
      set("calc") ((x, y) => sum(x, y)),
      set("upper") (x => toUpper(x)),
      set("Error") (m => throw(m)),
      set("composedFunctionToUpper") (
        () => get("CreateFunctionByJson") (get("MockedFunctionJson"))
      ),
      set("data") ([ "alpha", "beta", "omega" ]),
      set("MockedFunctionJson") (mock(map(toUpper))),
      set("CreateFunctionByJson") (json => build(json)),
      get("composedFunctionToUpper"), 
      fn => set("data")(fn(get("data"))),
      identity(get("upper")("example")), 
      x => of(toLower(x)),
      concat(get("data")),
      try (get("Error") ("TEST ERROR 1"), identity),
      set("acc", (acc, x) => concat(acc, of(toLower(x)))),
      reduce(get("acc"), []),
      map(x => toUpper(x)),
    )(x))
  `)
  ).toEqual(JSOF.parse(code1));
});

test("to match snapshot", () => {
  expect(JSOF.parse(code1)).toMatchSnapshot();
});
