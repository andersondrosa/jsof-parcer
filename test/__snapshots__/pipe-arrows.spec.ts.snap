// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`to match snapshot 1`] = `
{
  "_type": "caller",
  "args": [
    {
      "_type": "lambda",
      "args": [
        "x",
      ],
      "return": {
        "_type": "caller",
        "args": [
          {
            "_type": "prop",
            "path": "x",
          },
        ],
        "into": {
          "_type": "caller",
          "args": [
            {
              "_type": "caller",
              "args": [
                {
                  "_type": "lambda",
                  "args": [
                    "x",
                    "y",
                  ],
                  "return": {
                    "_type": "caller",
                    "args": [
                      {
                        "_type": "prop",
                        "path": "x",
                      },
                      {
                        "_type": "prop",
                        "path": "y",
                      },
                    ],
                    "into": {
                      "_type": "prop",
                      "path": "sum",
                    },
                  },
                },
              ],
              "into": {
                "_type": "caller",
                "args": [
                  "calc",
                ],
                "into": {
                  "_type": "prop",
                  "path": "set",
                },
              },
            },
            {
              "_type": "caller",
              "args": [
                {
                  "_type": "lambda",
                  "args": [
                    "x",
                  ],
                  "return": {
                    "_type": "caller",
                    "args": [
                      {
                        "_type": "prop",
                        "path": "x",
                      },
                    ],
                    "into": {
                      "_type": "prop",
                      "path": "toUpper",
                    },
                  },
                },
              ],
              "into": {
                "_type": "caller",
                "args": [
                  "upper",
                ],
                "into": {
                  "_type": "prop",
                  "path": "set",
                },
              },
            },
            {
              "_type": "caller",
              "args": [
                {
                  "_type": "lambda",
                  "args": [
                    "m",
                  ],
                  "return": {
                    "_type": "caller",
                    "args": [
                      {
                        "_type": "prop",
                        "path": "m",
                      },
                    ],
                    "into": {
                      "_type": "prop",
                      "path": "throw",
                    },
                  },
                },
              ],
              "into": {
                "_type": "caller",
                "args": [
                  "Error",
                ],
                "into": {
                  "_type": "prop",
                  "path": "set",
                },
              },
            },
            {
              "_type": "caller",
              "args": [
                {
                  "_type": "lambda",
                  "args": [],
                  "return": {
                    "_type": "caller",
                    "args": [
                      {
                        "_type": "caller",
                        "args": [
                          "MockedFunctionJson",
                        ],
                        "into": {
                          "_type": "prop",
                          "path": "get",
                        },
                      },
                    ],
                    "into": {
                      "_type": "caller",
                      "args": [
                        "CreateFunctionByJson",
                      ],
                      "into": {
                        "_type": "prop",
                        "path": "get",
                      },
                    },
                  },
                },
              ],
              "into": {
                "_type": "caller",
                "args": [
                  "composedFunctionToUpper",
                ],
                "into": {
                  "_type": "prop",
                  "path": "set",
                },
              },
            },
            {
              "_type": "caller",
              "args": [
                [
                  "alpha",
                  "beta",
                  "omega",
                ],
              ],
              "into": {
                "_type": "caller",
                "args": [
                  "data",
                ],
                "into": {
                  "_type": "prop",
                  "path": "set",
                },
              },
            },
            {
              "_type": "caller",
              "args": [
                {
                  "_type": "caller",
                  "args": [
                    {
                      "_type": "caller",
                      "args": [
                        {
                          "_type": "prop",
                          "path": "toUpper",
                        },
                      ],
                      "into": {
                        "_type": "prop",
                        "path": "map",
                      },
                    },
                  ],
                  "into": {
                    "_type": "prop",
                    "path": "mock",
                  },
                },
              ],
              "into": {
                "_type": "caller",
                "args": [
                  "MockedFunctionJson",
                ],
                "into": {
                  "_type": "prop",
                  "path": "set",
                },
              },
            },
            {
              "_type": "caller",
              "args": [
                {
                  "_type": "lambda",
                  "args": [
                    "json",
                  ],
                  "return": {
                    "_type": "caller",
                    "args": [
                      {
                        "_type": "prop",
                        "path": "json",
                      },
                    ],
                    "into": {
                      "_type": "prop",
                      "path": "build",
                    },
                  },
                },
              ],
              "into": {
                "_type": "caller",
                "args": [
                  "CreateFunctionByJson",
                ],
                "into": {
                  "_type": "prop",
                  "path": "set",
                },
              },
            },
            {
              "_type": "caller",
              "args": [
                "composedFunctionToUpper",
              ],
              "into": {
                "_type": "prop",
                "path": "get",
              },
            },
            {
              "_type": "lambda",
              "args": [
                "fn",
              ],
              "return": {
                "_type": "caller",
                "args": [
                  {
                    "_type": "caller",
                    "args": [
                      {
                        "_type": "caller",
                        "args": [
                          "data",
                        ],
                        "into": {
                          "_type": "prop",
                          "path": "get",
                        },
                      },
                    ],
                    "into": {
                      "_type": "prop",
                      "path": "fn",
                    },
                  },
                ],
                "into": {
                  "_type": "caller",
                  "args": [
                    "data",
                  ],
                  "into": {
                    "_type": "prop",
                    "path": "set",
                  },
                },
              },
            },
            {
              "_type": "caller",
              "args": [
                {
                  "_type": "caller",
                  "args": [
                    "example",
                  ],
                  "into": {
                    "_type": "caller",
                    "args": [
                      "upper",
                    ],
                    "into": {
                      "_type": "prop",
                      "path": "get",
                    },
                  },
                },
              ],
              "into": {
                "_type": "prop",
                "path": "identity",
              },
            },
            {
              "_type": "lambda",
              "args": [
                "x",
              ],
              "return": {
                "_type": "caller",
                "args": [
                  {
                    "_type": "caller",
                    "args": [
                      {
                        "_type": "prop",
                        "path": "x",
                      },
                    ],
                    "into": {
                      "_type": "prop",
                      "path": "toLower",
                    },
                  },
                ],
                "into": {
                  "_type": "prop",
                  "path": "of",
                },
              },
            },
            {
              "_type": "caller",
              "args": [
                {
                  "_type": "caller",
                  "args": [
                    "data",
                  ],
                  "into": {
                    "_type": "prop",
                    "path": "get",
                  },
                },
              ],
              "into": {
                "_type": "prop",
                "path": "concat",
              },
            },
            {
              "_type": "caller",
              "args": [
                {
                  "_type": "caller",
                  "args": [
                    "TEST ERROR 1",
                  ],
                  "into": {
                    "_type": "caller",
                    "args": [
                      "Error",
                    ],
                    "into": {
                      "_type": "prop",
                      "path": "get",
                    },
                  },
                },
                {
                  "_type": "prop",
                  "path": "identity",
                },
              ],
              "into": {
                "_type": "prop",
                "path": "try",
              },
            },
            {
              "_type": "caller",
              "args": [
                "acc",
                {
                  "_type": "lambda",
                  "args": [
                    "acc",
                    "x",
                  ],
                  "return": {
                    "_type": "caller",
                    "args": [
                      {
                        "_type": "prop",
                        "path": "acc",
                      },
                      {
                        "_type": "caller",
                        "args": [
                          {
                            "_type": "caller",
                            "args": [
                              {
                                "_type": "prop",
                                "path": "x",
                              },
                            ],
                            "into": {
                              "_type": "prop",
                              "path": "toLower",
                            },
                          },
                        ],
                        "into": {
                          "_type": "prop",
                          "path": "of",
                        },
                      },
                    ],
                    "into": {
                      "_type": "prop",
                      "path": "concat",
                    },
                  },
                },
              ],
              "into": {
                "_type": "prop",
                "path": "set",
              },
            },
            {
              "_type": "caller",
              "args": [
                {
                  "_type": "caller",
                  "args": [
                    "acc",
                  ],
                  "into": {
                    "_type": "prop",
                    "path": "get",
                  },
                },
                [],
              ],
              "into": {
                "_type": "prop",
                "path": "reduce",
              },
            },
            {
              "_type": "caller",
              "args": [
                {
                  "_type": "lambda",
                  "args": [
                    "x",
                  ],
                  "return": {
                    "_type": "caller",
                    "args": [
                      {
                        "_type": "prop",
                        "path": "x",
                      },
                    ],
                    "into": {
                      "_type": "prop",
                      "path": "toUpper",
                    },
                  },
                },
              ],
              "into": {
                "_type": "prop",
                "path": "map",
              },
            },
          ],
          "into": {
            "_type": "prop",
            "path": "pipe",
          },
        },
      },
    },
  ],
  "into": {
    "_type": "prop",
    "path": "context",
  },
}
`;
